elastic/kibana
# APM execution context - app, page, entitiy id

## Summary

This PR introduces a client side `ExecutionContextService` and uses it to provide application level context and passes it to APM transactions both on the client and the server side.

The information will be available for all APM RUM transactions as well as all server side APM request transactions.

All APM transactions are now automatically provided with: 
 * `name` - as provided by the `application` service

And Dashboard, Discover, Lens, Visualizations Editor and Dev Tools also provide:
 * `page` - a sub-division that can be used to identify an application's logical unit. For example for dashboards that would be `app` which is a single dashboard and `list` for the dashboards list.
 * `id` - a saved object id that can be used to identify a key entity. For example the dashboard ID for a dashboard or a saved search id for Discover. 

### How to test this PR?
* Enable APM by running `export ELASTIC_APM_ACTIVE=true`
* Run Kibana locally
* Turn off any ad blockers (they block FE events)
* Open [this view](https://kibana-cloud-apm.elastic.dev/app/discover) and load a saved search named `[Liza] localhost transactions`.
* Reenable and modify the `kibana_uuid` in the filter to your local `kibana_uuid`. (Some people have their uuid statically set by ftr tests to `5b2de169-2785-441b-ae8c-186a1936b17d` - if that's your id, delete your `.uuid` file.
* Navigate through your local kibana and verify that any frontend (`kibana-frontend`) and backend (`kibana`) events have all the required fields.

#### Home
![image](https://user-images.githubusercontent.com/3016806/154997752-0b75b2ca-0f28-4531-abc7-4e3b16665e82.png)

#### Discover
Saved
![image](https://user-images.githubusercontent.com/3016806/154994473-2f355362-62c1-4e3e-a67d-7eabf1f5c3e3.png)
Unsaved
![image](https://user-images.githubusercontent.com/3016806/154994402-d30e9f47-0eab-4410-8533-6fac278ff150.png)

#### Visualizations list
![image](https://user-images.githubusercontent.com/3016806/154997832-22bf24e3-46b9-4a37-ad3a-7d5ed98df44c.png)

#### Lens visualization
Unsaved
![image](https://user-images.githubusercontent.com/3016806/154994680-19e1bc4a-c2c5-4e41-a753-69197ffc8335.png)
Saved
![image](https://user-images.githubusercontent.com/3016806/154994731-7a45f7dc-8c5e-4fe1-b159-dab43bef72be.png)

#### Other visualizations
TSVB
![image](https://user-images.githubusercontent.com/3016806/155318570-359c92d7-ac62-43f3-ade4-931e2c6c3150.png)
Vega
![image](https://user-images.githubusercontent.com/3016806/155318457-fbfa6886-478c-4aa2-bd8f-e437d15eecbc.png)
Data Table
![image](https://user-images.githubusercontent.com/3016806/155319170-0d879340-c1e6-4a0c-af26-6a9855bbd478.png)
Timelion
![image](https://user-images.githubusercontent.com/3016806/155319253-14704185-8e41-4ccf-844e-e1cc81fecff0.png)


#### Dashboard list
![image](https://user-images.githubusercontent.com/3016806/154993931-818279e9-cd50-4936-b39a-eb4ef854ffe2.png)

#### Single dashboard 
Saved
![image](https://user-images.githubusercontent.com/3016806/154993723-eae9ab48-bc5a-4a47-ad1f-2f693abcd0a9.png)
Unsaved
![image](https://user-images.githubusercontent.com/3016806/154994295-acec770a-b6e2-45f8-a450-b82c572fb2fc.png)

#### Maps list
![image](https://user-images.githubusercontent.com/3016806/155004540-883d4a16-17ba-4e70-9f64-8ba38fd75c0b.png)

#### Single map
Unsaved
![image](https://user-images.githubusercontent.com/3016806/155004607-7e8648ca-1d5a-4cb2-a7f4-d2a810821fe7.png)
Saved
![image](https://user-images.githubusercontent.com/3016806/155004679-9e61da55-77cb-4729-8544-15de18f62a3d.png)



Here's a list of issues I'm already aware of:
* Bundles are loaded without any context
* `app-change` transactions get the wrong name

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
